### PR TITLE
Pipes - Lindsey Deuel - GroceryStore

### PR DESCRIPTION
**Pull request by VirtualLindsey**
# Grocery Store
Congratulations! You're submitting your assignment.

## Comprehension Questions

|  Question 	|   Response	|
|---	|---	|
|   Why is it useful to put classes inside modules? 	|   To be able to access methods in different classes easily	|
|   What is accomplished with `raise ArgumentError`? 	|  It stops the program if an argument isn't correct  	|
|   Why do you think we made the `.all` & `.find` methods class methods?  Why not instance methods?	|   Because you want to be able to see all instances of the classes, not look inside a given instance	|
|   Why does it make sense to use inheritance for the online order?  | Because a lot of the work is duplicated and DRY code is good code |
|   Did the presence of automated tests change the way you thought about the problem? How? | Yes. I looked or more opportunities to handle edge cases gracefully. |

### CODE REVIEW

File: specs/order_spec.rb [hunk 1]
```diff
@@ -79,31 +79,78 @@
...
+	it "Confirms the number of orders is correct"  do 
+		orders = Grocery::Order.all
+		rawCount = 0
+		CSV.foreach('../support/orders.csv') { |row|
+			rawCount += 1
+		}
+		
+		orderCount = 0
+		orders.each { |i|
+			orderCount += 1
```
Comment (droberts-sea) [inline specs/order_spec.rb:105]: Why not use `orders.length` here?

File: specs/order_spec.rb [hunk 2]
```diff
@@ -79,31 +79,78 @@
...
+	end
   end
 
   describe "Order.find" do
     it "Can find the first order from the CSV" do
-      # TODO: Your test code here!
+		myOrders = []
+      CSV.foreach('../support/orders.csv') { |row|
+				order_prodcuts = {}
+				order_id = row[0] #this is the ID
```
Comment (droberts-sea) [inline specs/order_spec.rb:117]: I'm curious why you included this big loop to read all the orders here. You put them all into the `myOrders` array, but as far as I can tell it isn't ever used.

The last two lines of this test seem to do everything you need (trying to find the first one and checking that the ID is as expected).

File: specs/online_order_spec.rb [hunk 3]
```diff
@@ -3,69 +3,137 @@
...
-
+      id = 1
+	  products = {'socks' => 2.34}
+	  customer = 5
+	  status = 'paid'
       # Instatiate your OnlineOrder here
-      # online_order =
-      # online_order.must_be_kind_of Grocery::Order
+      online_order = Grocery::OnlineOrder.new(id, products, customer, status)
+      online_order.must_be_kind_of Grocery::OnlineOrder
```
Comment (droberts-sea) [inline specs/online_order_spec.rb:23]: The intention here was to have a line like

```ruby
online_order.must_be_kind_of Grocery::Order
```

Which will verify the inheritance between `OnlineOrder` and `Order`

File: specs/online_order_spec.rb [hunk 4]
```diff
@@ -3,69 +3,137 @@
...
+		total = online_order.total
+		
+		total.must_equal 0
     end
   end
 
   describe "#add_product" do
     it "Does not permit action for processing, shipped or completed statuses" do
-      # TODO: Your test code here!
+		id = 1
```
Comment (droberts-sea) [inline specs/online_order_spec.rb:80]: This test doesn't actually run the `add_product` method! You should call that method and check that it either raises an error or returns a value to indicate it didn't work.

File: specs/online_order_spec.rb [hunk 5]
```diff
@@ -3,69 +3,137 @@
...
+		online_orders.length.must_equal raw_orders.length
+	end
   end
   
   describe "OnlineOrder.find_by_customer" do
     it "Returns an array of online orders for a specific customer ID" do
-      # TODO: Your test code here!
-    end
+		a = Grocery::OnlineOrder.find_by_customer(1)
+		a.must_be_kind_of Grocery::Customer
```
Comment (droberts-sea) [inline specs/online_order_spec.rb:136]: Based on the name of this test, shouldn't you be checking that `a` is of type `Array`, and the contents are instances of `Grocery::OnlineOrder`? If that's not what this method does, you should change the name of the test.

File: lib/onlineorder.rb [hunk 6]
```diff
@@ -0,0 +1,71 @@
+require 'csv'
+require_relative 'order.rb'
+require_relative 'customer.rb'
+
+module Grocery
+	class OnlineOrder < Order
+		attr_reader :id, :products, :customer, :status
+
+		def initialize(id, products, customer, status='pending')
+			super(id, products) #call the Order's constructor and passes variables as needed
```
Comment (droberts-sea) [inline lib/onlineorder.rb:10]: Good work with this class - you've done a really nice job of using inheritance and `super` to DRY up this code.

File: lib/order.rb [hunk 7]
```diff
@@ -1,18 +1,57 @@
...
+		def self.all()
+			myOrders = []
+			#row[0] = id | row[1] = orders
+			CSV.foreach('../support/orders.csv') { |row|
+				order_prodcuts = {}
+				order_id = row[0] #this is the ID
+				row[1].split(';').each { |i| #this splits the order string into individual name/cost pairs
+					single_item = i.split(':')#separates the name from the cost
+					cost = single_item[1]
+					name = single_item[0]
```
Comment (droberts-sea) [inline lib/order.rb:22]: I like the logic you've written to parse the list of products. Your local variables (`single_item`, `cost` and `name`) are clearly named and make it easy to tell what's going on.

File: lib/onlineorder.rb [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+				if i.id.to_i == id.to_i
+					return i
+				end
+			}
+			raise ArgumentError, "There is no Order with the ID: #{id}" 
+		end
+		
+		def self.find_by_customer(customer_id)
+			customers = Grocery::Customer.all
+			customers.each { |i|
```
Comment (droberts-sea) [inline lib/onlineorder.rb:45]: The idea for this method was to return a list of all the `OnlineOrders` matching this customer ID. Finding a customer by ID is already handled by `Customer.find`.

File: specs/order_spec.rb [hunk 9]
```diff
@@ -79,31 +79,78 @@
...
+		order = Grocery::Order.find(100)
+		
+		order.id.to_i.must_equal 100
+	  
+	  
     end
 
     it "Raises an error for an order that doesn't exist" do
-      # TODO: Your test code here!
+      proc {Grocery::Order.find(1235)}.must_raise ArgumentError
```
Comment (droberts-sea) [inline specs/order_spec.rb:153]: Nice - very clear and concise.

**Comment by droberts-sea** [comment]
# Grocery Store
## What We're Looking For

|  Feature  |   Feedback|
|---	|---	|
|  **Baseline** |   |
|   Answered comprehension questions	|  yes  |
|   Used Git Regularly	|  no - Please make an effort to practice good git habits on these projects. Making this a part of your regular workflow is very valuable, and it'll be much easier to do now than to force it in later. If you're having trouble remembering to pause what you're doing and commit, [here is an article that might be worth a read](https://dev.to/rpalo/plan-your-commits).  |
|  **Wave 1** |   |
|   All provided tests pass | yes |
|   Using the appropriate `attr_` for instance variables |  yes  |
|  **Wave 2** |   |
|   All stubbed tests are implemented fully and pass | yes |
|   Appropriately parses the product data from CSV file in `Order.all` | yes |
|   Used `CSV` library only in `Order.all` (not in `Order.find`) |  yes |
|   Used `Order.all` to get account list in `Order.find`	|  yes  |
|  **Wave 3** |   |
|   All stubbed tests are implemented fully and pass | yes |
|   Used inheritance in the initialize for online order |  yes  |
|   Used inheritance for the total method in online order | yes |
|   Appropriately created Customer class and class methods `all` & `find` | yes |
|   **Additional Notes**	|   	|

I notice that some of your files are using tabs for indentation instead of spaces. There's nothing especially wrong with tabs, but it's important for a group of people working together to all use the same system - otherwise things will get very messy very fast. This becomes especially true once you start to work on group projects and share code between computers. In this class we'll stick with using 2 spaces for indentation. Please set up your editor to follow this convention, and let me know if you need help wading through the options.

This looks good, especially the code in `lib/`. It seems like there was a requirements miss on `OnlineOrder.find_by_customer`, and I've identified a few places below where your tests could be clarified somewhat, but in general it's a very solid submission. Keep up the good work!
